ddtrace/contrib/wsgi/wsgi.py

```diff
@@ -0,0 +1,159 @@
+import sys
+
+from ddtrace.compat import PY2
+
+if PY2:
+    import exceptions
+
+    generatorExit = exceptions.GeneratorExit
+else:
+    import builtins
+
+    generatorExit = builtins.GeneratorExit
+
+
+from ddtrace.vendor import six
+
+import ddtrace
+from ddtrace import config
+from ddtrace.ext import SpanTypes
+from ddtrace.internal.logger import get_logger
+from ddtrace.propagation.http import HTTPPropagator
+from ddtrace.propagation.utils import from_wsgi_header
+
+from six.moves.urllib.parse import quote
+
+from .. import trace_utils
+
+
+log = get_logger(__name__)
+
+propagator = HTTPPropagator()
+
+config._add(
+    "wsgi",
+    dict(
+        _default_service="wsgi",
+        distributed_tracing=True,
+    ),
+)
+
+
+def construct_url(environ):
+    """
+    https://www.python.org/dev/peps/pep-3333/#url-reconstruction
+    """
+    url = environ["wsgi.url_scheme"] + "://"
+
+    if environ.get("HTTP_HOST"):
+        url += environ["HTTP_HOST"]
+    else:
+        url += environ["SERVER_NAME"]
+
+        if environ["wsgi.url_scheme"] == "https":
+            if environ["SERVER_PORT"] != "443":
+                url += ":" + environ["SERVER_PORT"]
+        else:
+            if environ["SERVER_PORT"] != "80":
+                url += ":" + environ["SERVER_PORT"]
+
+    url += quote(environ.get("SCRIPT_NAME", ""))
+    url += quote(environ.get("PATH_INFO", ""))
+    if environ.get("QUERY_STRING"):
+        url += "?" + environ["QUERY_STRING"]
+
+    return url
+
+
+def get_request_headers(environ):
+    """
+    Manually grab the request headers from the environ dictionary.
+    """
+    request_headers = dict()
```

Comment:
```suggestion
    request_headers = {}
```

lol for some reason this is 5x faster

```
In [1]: %timeit dict()
107 ns ± 4.67 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [2]: %timeit {}
20.2 ns ± 1.14 ns per loop (mean ± std. dev. of 7 runs, 100000000 loops each)
```